backend/dataall/modules/dataset_sharing/services/share_managers/s3_access_point_share_manager.py

```diff
@@ -356,31 +387,52 @@ def delete_target_role_access_policy(
         logger.info(
             'Deleting target role IAM policy...'
         )
-        access_point_name = S3ShareManager.build_access_point_name(share)
+        access_point_name = S3AccessPointShareManager.build_access_point_name(share)
         existing_policy = IAM.get_role_policy(
             target_environment.AwsAccountId,
             share.principalIAMRoleName,
-            "targetDatasetAccessControlPolicy",
+            IAM_ACCESS_POINT_ROLE_POLICY,
         )
+        key_alias = f"alias/{dataset.KmsAlias}"
+        kms_client = KmsClient(dataset.AwsAccountId, dataset.region)
+        kms_key_id = kms_client.get_key_id(key_alias)
         if existing_policy:
-            if dataset.S3BucketName in ",".join(existing_policy["Statement"][0]["Resource"]):
-                target_resources = [
-                    f"arn:aws:s3:::{dataset.S3BucketName}",
-                    f"arn:aws:s3:::{dataset.S3BucketName}/*",
-                    f"arn:aws:s3:{dataset.region}:{dataset.AwsAccountId}:accesspoint/{access_point_name}",
-                    f"arn:aws:s3:{dataset.region}:{dataset.AwsAccountId}:accesspoint/{access_point_name}/*"
-                ]
-                for item in target_resources:
-                    existing_policy["Statement"][0]["Resource"].remove(item)
-                if not existing_policy["Statement"][0]["Resource"]:
-                    IAM.delete_role_policy(target_environment.AwsAccountId, share.principalIAMRoleName, "targetDatasetAccessControlPolicy")
-                else:
-                    IAM.update_role_policy(
-                        target_environment.AwsAccountId,
-                        share.principalIAMRoleName,
-                        "targetDatasetAccessControlPolicy",
-                        json.dumps(existing_policy),
-                    )
+            s3_target_resources = [
+                f"arn:aws:s3:::{dataset.S3BucketName}",
+                f"arn:aws:s3:::{dataset.S3BucketName}/*",
+                f"arn:aws:s3:{dataset.region}:{dataset.AwsAccountId}:accesspoint/{access_point_name}",
+                f"arn:aws:s3:{dataset.region}:{dataset.AwsAccountId}:accesspoint/{access_point_name}/*"
+            ]
+            ShareManagerUtils.remove_resource_from_statement(
+                existing_policy["Statement"][0],
+                s3_target_resources
+            )
+
+            kms_target_resources = [
+                f"arn:aws:kms:{dataset.region}:{dataset.AwsAccountId}:key/{kms_key_id}",
+                f"arn:aws:kms:{dataset.region}:{dataset.AwsAccountId}:key/{kms_key_id}/*"
```

Comment:
Unless I am mistaken - the kms target resource `f"arn:aws:kms:{dataset.region}:{dataset.AwsAccountId}:key/{kms_key_id}/*"` does not mean anything to IAM permissions. This type of formatting only applies to S3 Buckets and their object contents - having the S3Arn + the S3Arn`/*`

I think we should be okay to remove the second `kms_target_resource` that includes the trailing `/*` everywhere

Reply:
Good catch!